Update primer octicons and add support for all sizes

jacortinas opened the pull request:
Related to https://github.com/desktop/desktop/pull/18035

## Description
This PR updates the [`@primer/octicons`](https://www.npmjs.com/package/@primer/octicons) package to the most recent version. This allows access to new icons such as accessibility, copilot, etc. This also keep us more in sync with the current dotcom look and feel.

One of the key differences between our existing version and new versions of icons is that the root `<svg>` element can now have multiple nested `<path>` elements. The `script/generate-octicons.ts` script that desktop uses to generate a typed list of icons only currently supports pulling the path data from the first path element of the icon. This is why the `copy` icon among others was rendered incorrectly after updating the library.

![Screenshot of an incorrectly rendered copy icon](https://github.com/desktop/desktop/assets/171215/af4f5c25-f2f6-4b94-a6d1-a17f98644e24)

In order to support multiple paths, the `generate-octicons.ts` script has been updated to use the newly available `ast` object of each icon, which contains all of the information necessary to recreate the `<svg>` tag and all child tags without requiring any HTML or XML parsing. The structure of the generated data has been adjusted to include all sizes and all call sites and usages of it have been adjusted as well.

The copy icon was previously only available at 24px, which predicated making all provided icon sizes available in the generated output. This now enables a new `height` property on the `Octicon` component that can choose the best fitting SVG paths.

Refactoring Octicons:

* All instances of `import * as OcticonSymbol from '../octicons/octicons.generated'` have been replaced with `import * as octicons from '../octicons/octicons.generated'` across various files. This change standardizes the import name for Octicons across the codebase. [[1]](diffhunk://#diff-3c7407bceff5a216d1f5d7d7f012cd0cd453b4e402cc642b5e26d885aa1b1180L7-R7) [[2]](diffhunk://#diff-43fa98f0d8738d64c9b6faac4ec5d1fe52a3eb0256028c4ffcd1b1335f96804cL27-R27) [[3]](diffhunk://#diff-fee69b20e1122c907cf2cec44742d772303ad26f0eb466ccba629bcbc17f84acL5-R5) [[4]](diffhunk://#diff-b952c5a52a0301efa2a709b19de65d9274e0b3837c42bbac611f1c9ef871f8d0L54-R55) [[5]](diffhunk://#diff-d9845e9610812a950af360df674dbb375f32bc1d9073bcbd8a1c40a7dcd0db1cL3-R3) [[6]](diffhunk://#diff-c5bd290b405efd75fae83a2552fa9f4ef2aaad87ce2e0e167c5171e5bf2e38fdL3-R3) [[7]](diffhunk://#diff-d92e6c09246cb2f30aad5fb6a59ba67628112a7f4603665a6e63372e18bdbab9L3-R3) [[8]](diffhunk://#diff-8aed8fb53ff0679d81adbaa5bf314a127b06ea62d83cd1088f687e16ff3d4a79L3-R3) [[9]](diffhunk://#diff-4d672515d4da0441268d815c59fb790c25310e68c156dc02ffbceaacc29be860L3-R3) [[10]](diffhunk://#diff-6237fd1e72628abe3162d17ff8f6b3b3b7e77e075b84f47f24cc2c0fe9440721L3-R3) [[11]](diffhunk://#diff-dc8e5641d19dfd84c7c0831d83413f8ec8fc10d796f647651874b7861c706b36L4-R4) [[12]](diffhunk://#diff-9bff33c2d9a342a61b1cfe3d68f3ae020a0060da9c5c31dc0aeb8d3cf6eca2f0L10-R10) [[13]](diffhunk://#diff-5dcedc872c02ab6ff20ec5be0f0ecb81ddfe1662b0e5c301f39fe12c09cbd976L3-R3) [[14]](diffhunk://#diff-b077bcf646fa44635e0ede6116e1d3599ca23db65d48e09b6a5eff4078fc3bd5L6-R6) [[15]](diffhunk://#diff-07f7438f7d9e25fd1d32a1ce73e3f76ab37201d38430b8be86199c5fdf4bc727L20-R20)

* All instances where Octicons were used in the code, the reference `OcticonSymbol` has been replaced with `octicons`. [[1]](diffhunk://#diff-3c7407bceff5a216d1f5d7d7f012cd0cd453b4e402cc642b5e26d885aa1b1180L144-R144) [[2]](diffhunk://#diff-43fa98f0d8738d64c9b6faac4ec5d1fe52a3eb0256028c4ffcd1b1335f96804cL446-R446) [[3]](diffhunk://#diff-fee69b20e1122c907cf2cec44742d772303ad26f0eb466ccba629bcbc17f84acL98-R100) [[4]](diffhunk://#diff-fee69b20e1122c907cf2cec44742d772303ad26f0eb466ccba629bcbc17f84acL156-R156) [[5]](diffhunk://#diff-b952c5a52a0301efa2a709b19de65d9274e0b3837c42bbac611f1c9ef871f8d0L2885-R2895) [[6]](diffhunk://#diff-d9845e9610812a950af360df674dbb375f32bc1d9073bcbd8a1c40a7dcd0db1cL40-R40) [[7]](diffhunk://#diff-c5bd290b405efd75fae83a2552fa9f4ef2aaad87ce2e0e167c5171e5bf2e38fdL32-R32) [[8]](diffhunk://#diff-d92e6c09246cb2f30aad5fb6a59ba67628112a7f4603665a6e63372e18bdbab9L38-R38) [[9]](diffhunk://#diff-8aed8fb53ff0679d81adbaa5bf314a127b06ea62d83cd1088f687e16ff3d4a79L44-R44) [[10]](diffhunk://#diff-4d672515d4da0441268d815c59fb790c25310e68c156dc02ffbceaacc29be860L34-R34) [[11]](diffhunk://#diff-6237fd1e72628abe3162d17ff8f6b3b3b7e77e075b84f47f24cc2c0fe9440721L41-R41) [[12]](diffhunk://#diff-dc8e5641d19dfd84c7c0831d83413f8ec8fc10d796f647651874b7861c706b36L39-R39) [[13]](diffhunk://#diff-9bff33c2d9a342a61b1cfe3d68f3ae020a0060da9c5c31dc0aeb8d3cf6eca2f0L42-R42) [[14]](diffhunk://#diff-5dcedc872c02ab6ff20ec5be0f0ecb81ddfe1662b0e5c301f39fe12c09cbd976L26-R26) [[15]](diffhunk://#diff-b077bcf646fa44635e0ede6116e1d3599ca23db65d48e09b6a5eff4078fc3bd5L95-R95) [[16]](diffhunk://#diff-07f7438f7d9e25fd1d32a1ce73e3f76ab37201d38430b8be86199c5fdf4bc727L172-R172) [[17]](diffhunk://#diff-07f7438f7d9e25fd1d32a1ce73e3f76ab37201d38430b8be86199c5fdf4bc727L297-R297)

ESLint changes:

* In `.eslintrc.yml`, a new rule `@typescript-eslint/naming-convention: off` has been added specifically for the file `app/src/ui/octicons/octicons.generated.ts`. This change disables the naming convention rule for this particular file.

### Screenshots

> Previous publish branch and copy icons above, new icons below.
> ![Screenshot comparing old and new icons](https://github.com/desktop/desktop/assets/171215/985f528f-a6db-4c00-84cb-6bac6689278e)

> Preferences icons look fine, new gear icon is slightly stronger from previous icon.
> ![Screenshot of preferences dialog icons](https://github.com/desktop/desktop/assets/171215/afd9b386-c335-4463-8289-6ea9f6161e5e)

> Submodule changes screen
> <img width="678" alt="Screenshot of submodule changes screen" src="https://github.com/desktop/desktop/assets/171215/ee16919f-71c0-4214-a3f5-d842a8b93604">

> Commit list branch icon
> <img width="63" alt="Screenshot of commit list branch icon" src="https://github.com/desktop/desktop/assets/171215/48816d68-b756-4f55-8f40-a2bc86072f91">

> File change icons and new diff checkmarks
> <img width="98" alt="Screenshot of file change icons and new diff checkmarks" src="https://github.com/desktop/desktop/assets/171215/f9864f36-402c-45f4-bab1-c54ff4615223">

> Missing newline EOF icon
> <img width="131" alt="Screenshot of missing newline icon" src="https://github.com/desktop/desktop/assets/171215/cd624043-650f-4905-b57e-89cf7ec835a0">

> Stashed changes icon
> ![Screenshot of stashed changes icon](https://github.com/desktop/desktop/assets/171215/96725396-c73e-4394-b2ae-ecbe9033aa63)

> Checkmark filled circle icon
> ![Screenshot of check filled circle icon](https://github.com/desktop/desktop/assets/171215/fffb73e3-e7de-440c-9e07-634fdfe69039)

> Co-Authors icon
> ![Screenshot of co-authors icon](https://github.com/desktop/desktop/assets/171215/591045ae-f017-45bb-92c3-8371c9c9b2d8)


## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes:

jacortinas committed:
Update primer/octicons package

jacortinas committed:
Update generate script to support multiple paths

The latest releases of the octicons package now have an ast object available, we use that to get the path data for the icons now instead of parsing the path elements.

jacortinas committed:
Update generated icons and lint rules

jacortinas committed:
Support multiple paths in Octicon component render

jacortinas committed:
Adjust manually created icons to new type

jacortinas committed:
settings -> gear

jacortinas committed:
Scale down success check and copy icon

jacortinas committed:
Use checkCircleFill icon for success

jacortinas committed:
Resize all copy buttons to 12px

jacortinas committed:
Rewrite icon generation again

- All paths and size values are included
- Symbol name is included
- More strict type checking is enabled by enumerating all of the symbol names and sizes and defining types.

jacortinas committed:
Regenerate static icons and types

jacortinas committed:
Update Octicon component for new data struct

jacortinas committed:
Update all call sites and custom icon defs

jacortinas committed:
Move Custom type to index

jacortinas committed:
Use v for heights and less info for custom icons

tidy-dev reviewed (changes requested):
This makes sense to me and the added flexibility is really nice.  Plus updating and being able to update the Octicon dependency is great. 


I have a couple questions.. but more stylistic in nature. But, I think we can simplify the code a little to make easier on future devs looking to understand this. I am probably missing the use case, but mind elaborating on why the add of symbol name?

Looks like we went from:
```
export const alert:  OcticonSymbolType = {
  w: 16,
  h:  16,
  d: '....',
  fr: 'evenodd'
}
```

to 

```
export const alert: OcticonSymbolType = {
  s: "alert",
  v: {
    "16": {
      p: [ "..."],
      h: 16,
      w: 16,
    },
    "24": {
      p: [ "...", "...", ],
      h: 24,
      w: 24,
    },
  },
};
```

I follow the need for the map of sizes and array of paths. 

But, not sure I understand the need for the `s` and the `v`, the `s` appears to be meta data used in as a class name but not required? Could we instead make each accessibility constants = `v` like:
```
export const alert: OcticonSymbolType = {
    "16": {
      p: [ "..."],
      h: 16,
      w: 16,
    },
    "24": {
      p: [ "...", "...", ],
      h: 24,
      w: 24,
    },
};
```

On a similar vein of reducing complexity, it looks like the 
```
export type CustomOcticonSymbolType = {
  /** The symbol name */
  readonly s?: string

  /** The symbol width */
  readonly w: number

  /** The symbol height */
  readonly h: number

  /** The symbol SVG paths */
  readonly d: string | string[]
}
```

could be reuse the:
```
export type OcticonSymbolHeight<N extends OcticonSymbolSize> = {
  /** SVG path element data */
  readonly p: string[],

  /** The width of the symbol */
  readonly w: number,

  /** The height of the symbol */
  readonly h: N
}
```

By making our custom symbols have an array of one path. 


Other thoughts: For a PR like this, a nice-to-have, would be a pic of the usage of each unique `<Octicon` use, just make sure we didn't impact any of the existing use cases.

jacortinas committed:
Update icon generation and types

- Stop storing non-essential information like the name of icons.
- Reorganize type names and definitions to avoid redundant "Type" suffix.

jacortinas committed:
Update generated icon output

jacortinas committed:
Update Octicon component for simpler structure

jacortinas committed:
Clean up for new types

- Rename "* as OcticonSymbol" to "* as octicons" to denote a module namespace instead of a type.
- Change icon custom type to variant type.

jacortinas committed:
Fix import in generate-icons that broke everything

jacortinas commented:
@tidy-dev I've updated the types to remove any redundant nesting or data. I've also cleaned up the types to reuse whatever I could. I've taken over the `OcticonSymbol` as a type, so I changed the confusing variable we were importing into with the same name. The module name when we import all of the generated symbols is now `octicons`, and I've made those changes everywhere.

jacortinas committed:
Update a comment

tidy-dev commented on app/src/ui/octicons/octicon.tsx:
```diff
@@ -34,14 +36,34 @@ interface IOcticonProps {
+        height
+      )
+
+      const scaledSymbol = symbol[naturalHeight]!
```
```suggestion
      const scaledSymbol = symbol[naturalHeight]
      if(scaledSymbol === undefined) {
           // Just a type guard, shouldn't happen as we are getting the height from the symbol object.
          return null
      }
```

We avoid using the type assertion in favor of type guards since type assertion bypassing the purpose of using typescript. Type guards let us intentionally define how we want to handle the "should be impossible" case. Here, I think a comment and a return null is fine -> but others it might make sense to do a non fatal error to help debug. But, I think if we ever had an instance of a Octicon not showing up, we would look at the Octicon file. :P

tidy-dev commented on app/src/ui/octicons/sync-clockwise.ts:
```diff
@@ -1,17 +1,17 @@
  * the `sync` octicon.
  */
-export const syncClockwise: OcticonSymbolType = {
+export const syncClockwise: OcticonSymbolVariant = {
```
Curious, could we make the two custom octicons (this one and `diffCheck`) be `OcticonSymbolVariants` type and simplify our `octicon.tsx` code since we wouldn't need to check for single versus a map. Just always expect as map and these one's just happen to have 1 size in the map.

tidy-dev replied:
I see we have a few other scattered through the files too. :P I guess we do not have a very consistent pattern on where we store custom octicons. :P

jacortinas replied:
@tidy-dev Yeah I was going to say something similar. I think being able to define a single custom variant and pass that in with no need to nest the custom icon in another level of a hash is good compromise. I did go back and forth a couple of times.

tidy-dev commented on app/src/ui/octicons/octicons.generated.ts:
```diff
@@ -1,1579 +1,348 @@
- * This file is automatically generated by the generate-octicons tool.
- * Manually changing this file will only lead to sadness.
- */
+* This file is automatically generated by the generate-octicons tool.
```
I think prettier or some formatter was run on the last version of this file. Not terribly important but a nice to have and that way don't have to put an exception in the prettier file since this doesn't get generated very often.

jacortinas replied:
@tidy-dev With prettier formatting this file ends up at over 5000 lines. I think with one type definition per line the file ends up at a human readable 348 lines that are easy to scan through.

```typescript
export const zoomOut: OcticonSymbolVariants = {
  '16': {
    p: [
      'M4.5 6.75h6a.75.75 0 0 1 0 1.5h-6a.75.75 0 0 1 0-1.5Z',
      'M0 7.5a7.5 7.5 0 1 1 13.307 4.747l2.473 2.473a.749.749 0 1 1-1.06 1.06l-2.473-2.473A7.5 7.5 0 0 1 0 7.5Zm7.5-6a6 6 0 1 0 0 12 6 6 0 0 0 0-12Z',
    ],
    h: 16,
    w: 16,
  },
  '24': {
    p: [
      'M14.5 11.25a.75.75 0 0 0 0-1.5h-8a.75.75 0 0 0 0 1.5h8Z',
      'M0 10.5C0 4.701 4.701 0 10.5 0S21 4.701 21 10.5c0 2.63-.967 5.033-2.564 6.875l4.344 4.345a.749.749 0 1 1-1.06 1.06l-4.345-4.344A10.459 10.459 0 0 1 10.5 21C4.701 21 0 16.299 0 10.5Zm10.5-9a9 9 0 0 0-9 9 9 9 0 0 0 9 9 9 9 0 0 0 9-9 9 9 0 0 0-9-9Z',
    ],
    h: 24,
    w: 24,
  },
```

vs.

```typescript
export const zoomOut: OcticonSymbolVariants = {"16":{"p":["M4.5 6.75h6a.75.75 0 0 1 0 1.5h-6a.75.75 0 0 1 0-1.5Z","M0 7.5a7.5 7.5 0 1 1 13.307 4.747l2.473 2.473a.749.749 0 1 1-1.06 1.06l-2.473-2.473A7.5 7.5 0 0 1 0 7.5Zm7.5-6a6 6 0 1 0 0 12 6 6 0 0 0 0-12Z"],"h":16,"w":16},"24":{"p":["M14.5 11.25a.75.75 0 0 0 0-1.5h-8a.75.75 0 0 0 0 1.5h8Z","M0 10.5C0 4.701 4.701 0 10.5 0S21 4.701 21 10.5c0 2.63-.967 5.033-2.564 6.875l4.344 4.345a.749.749 0 1 1-1.06 1.06l-4.345-4.344A10.459 10.459 0 0 1 10.5 21C4.701 21 0 16.299 0 10.5Zm10.5-9a9 9 0 0 0-9 9 9 9 0 0 0 9 9 9 9 0 0 0 9-9 9 9 0 0 0-9-9Z"],"h":24,"w":24}}
```

tidy-dev replied:
Hmmm.. I am going back and forth. I think I still favor the formatted view because I can quickly see what sizes are available for a given Octicon (which if I was placing one and realized the size isn't what I needed, I may search this file just to see the definition of that Octicon). Secondly, having it formatted shows me what the type structure is for understanding how this code works at a glance. I don't really see this file as one I would scan for much. For example, if I needed to find a new Octicon, I typically look at the primer website were I can visually see the octicons to select from.

jacortinas committed:
Allow prettier to format generated icons file

jacortinas committed:
No need for a type assertion

jacortinas committed:
Add a type guard for the missing case

tidy-dev reviewed (approved):
✨  Thanks for those changes.
